skbio/stats/distance/_base.py

```diff
@@ -738,6 +740,118 @@ def from_iterable(cls, iterable, metric, key=None, keys=None):
 
         return cls(dm, keys_)
 
+    @classonlymethod
+    @experimental(as_of="0.4.2-dev")
+    def from_pairwise_distances(cls, pairwise_distances):
+        """Create ``DistanceMatrix`` from an iterable of ``PairwiseDistances``.
+
+        Constructs a ``DistanceMatrix`` from the complete set of IDs defined by
+        `pairwise_distances`. All "between" pairwise distances must be defined.
+
+        Parameters
+        ----------
+        pairwise_distances : iterable of PairwiseDistances
+            Pairwise distances to merge into a single ``DistanceMatrix``.
+
+        Returns
+        -------
+        DistanceMatrix
+            Distance matrix representing all pairwise distances between the
+            complete set of IDs present in `pairwise_distances`. The IDs in the
+            distance matrix will be in sorted order as there is no ordering
+            defined by `pairwise_distances`.
+
+        Raises
+        ------
+        TypeError
+            If `pairwise_distances` does not contain ``PairwiseDistances``
+            objects.
+        ValueError
+            If there are conflicting distances between a pair of IDs (e.g., two
+            ``PairwiseDistances`` objects define a different distance between
+            the same pair of IDs).
+        ValueError
+            If there is not at least one ID defined across
+            `pairwise_distances`.
+        ValueError
+            If any "between" distances are missing.
+
+        See Also
+        --------
+        skbio.stats.distance.PairwiseDistances
+        skbio.diversity.beta_diversity
+
+        Examples
+        --------
+        Construct a ``DistanceMatrix`` from two ``PairwiseDistances`` objects
+        collectively defining three IDs, `'A'`, `'B'`, and `'C'`:
+
+        >>> from skbio.stats.distance import DistanceMatrix, PairwiseDistances
+        >>> pwdists = [
+        ...     PairwiseDistances({('A', 'B'): 0.3, ('A', 'C'): 0.1}),
+        ...     PairwiseDistances({('B', 'C'): 0.2})
+        ... ]
+        >>> dm = DistanceMatrix.from_pairwise_distances(pwdists)
+        >>> dm
+        3x3 distance matrix
+        IDs:
+        'A', 'B', 'C'
+        Data:
+        [[ 0.   0.3  0.1]
+         [ 0.3  0.   0.2]
+         [ 0.1  0.2  0. ]]
+
+        Note that the IDs are in sorted order.
+
+        """
+        distances = {}
```

Comment:
Unfortunately, row-column-value using Python `dict`s will not scale to the size of datasets that we're developing #1337 for. Or at least, a mock test using 25k samples on my laptop killed it after consuming 16GB of memory. The fundamental issue is the memory allocation strategy that Python uses on dicts, and this parallels the RCV issues we had with BIOM. Would it be possible to instead populate a numpy array or sparse arrays? eg:

```python
distances = np.zeros([])
ids_in_distances = set()
for pwdist in pairwise_distances:
    # resize disances if necessary given IDs
    for i, j in pwdist:
        distances[i, j] = pwdist[(i, j)]
```

I'm wondering though if the simpliest thing is the best thing, were we just use dense DistanceMatrices at all blocks and then just provide a `DistanceMatrix.__add__`?